## PACKAGES

In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

## 1. IMPORTING DATA FROM SQL SERVER

### connecting to the database

In [5]:
from sqlalchemy import create_engine
from sqlalchemy.engine import URL

In [6]:
import psycopg2
hostname = 'localhost'
database = 'nyfd'
user_name = 'postgres'
pwd = 'raksha' 
port_id= 5432
conn=None
cur=None

In [7]:
engine = create_engine(f'postgresql://{user_name}:{pwd}@{hostname}:{port_id}/{database}')

In [8]:
sql = "SELECT * FROM public.insp"

In [9]:
insp = pd.read_sql_query(sql,engine)
insp.sample(10)

facility  \
1112890            JOEY B'S AT BRICKSTONE   
1049351       DUNKIN DONUTS CALLEN DONUTS   
134366          CALDA PIZZERIA RESTAURANT   
495512                        DAIRY QUEEN   
443641                   LARCHMONT TAVERN   
302502   Riverview Bar & Restaurant, Inc.   
1293758      WOODSIDE ACRES POPPY'S PLACE   
576975             FAT PATTIES BURGER BAR   
1159424                     DUNKIN DONUTS   
340903                          Fish Cove   

                                         address date_of_inspection  \
1112890          1325 ELMWOOD AVENUE,  ROCHESTER         12/20/2018   
1049351               34 UTICA STREET,  HAMILTON          1/28/2013   
134366     2890-3 HEMPSTEAD TURNPIKE,  LEVITTOWN         11/27/2023   
495512                     2319 ROUTE 66,  GHENT          2/26/2010   
443641         104 CHATSWORTH AVENUE,  LARCHMONT          2/14/2022   
302502     928 State Highway 11C,  Brasher Falls         12/22/2022   
1293758  225 MUTTONTOWN EASTWOODS ROAD,  SYOSSET          6/24/2015   
576975      1123 UPPER FRONT STREET,  BINGHAMTON          2/29/2016   
1159424       681 EAST MAIN STREET,  MOUNT KISCO           1/8/2013   
340903    458 SOUTH Main STREET,  North Syracuse         11/13/2009   

        violation_item                              violation_description  \
1112890           None                                               None   
1049351            15A  Floors, walls, ceilings, not smooth, properly ...   
134366             11C  Food contact surfaces not washed, rinsed and s...   
495512            None                                               None   
443641              8A  Food not protected during storage, preparation...   
302502             15A  Floors, walls, ceilings, not smooth, properly ...   
1293758           None                                               None   
576975             15B  Lighting and ventilation inadequate, fixtures ...   
1159424             8A  Food not protected during storage, preparation...   
340903              5A  Potentially hazardous foods are not kept at or...   

             critical_violation  total_critical_violations  \
1112890  Not Critical Violation                        0.0   
1049351  Not Critical Violation                        0.0   
134366   Not Critical Violation                        2.0   
495512   Not Critical Violation                        0.0   
443641   Not Critical Violation                        5.0   
302502   Not Critical Violation                        0.0   
1293758  Not Critical Violation                        0.0   
576975   Not Critical Violation                        0.0   
1159424  Not Critical Violation                        0.0   
340903       Critical Violation                        2.0   

         total_crit_not_corrected  total_noncritical_violations  \
1112890                       0.0                           0.0   
1049351                       0.0                           2.0   
134366                        0.0                           3.0   
495512                        0.0                           0.0   
443641                        0.0                          18.0   
302502                        0.0                           5.0   
1293758                       0.0                           0.0   
576975                        0.0                           4.0   
1159424                       0.0                           1.0   
340903                        0.0                           2.0   

        local_health_department  ... permitted_dba  \
1112890           Monroe County  ...          None   
1049351          Madison County  ...          None   
134366           Nassau  County  ...          None   
495512          Columbia County  ...          None   
443641       Westchester County  ...          None   
302502   Canton District Office  ...          None   
1293758          Nassau  County  ...          None   
576975            Broome County  ...          None   


In [10]:
insp.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1310873 entries, 0 to 1310872
Data columns (total 31 columns):
 #   Column                        Non-Null Count    Dtype  
---  ------                        --------------    -----  
 0   facility                      1310834 non-null  object 
 1   address                       1310832 non-null  object 
 2   date_of_inspection            1310834 non-null  object 
 3   violation_item                1069652 non-null  object 
 4   violation_description         1069652 non-null  object 
 5   critical_violation            1310834 non-null  object 
 6   total_critical_violations     1230072 non-null  float64
 7   total_crit_not_corrected      1229454 non-null  float64
 8   total_noncritical_violations  1230201 non-null  float64
 9   local_health_department       1310834 non-null  object 
 10  county                        1310834 non-null  object 
 11  facility_code                 1310834 non-null  object 
 12  facility_address            

- Overall there are null values in the dataset. 
-  Column `Date of inspection` datatype should be changed.

## 2. PRELIMINARY ANALYSIS

### 2.1 CHECKING DATATYPES

In [14]:
insp.sample(5)

facility                             address  \
692250  LITTLE KABAB STATION   31 EAST MAIN STREET,  MOUNT KISCO   
598737      Ice Cream Island     130  Franklin Plaza,  DANSVILLE   
960725     STADIUM GOLF CLUB    333 JACKSON AVENUE,  SCHENECTADY   
366251        18 NORTH GRILL       18 WESTAGE   DRIVE,  FISHKILL   
24645    HATTIE'S RESTAURANT  45 PHILA STREET,  SARATOGA SPRINGS   

       date_of_inspection violation_item  \
692250           2/4/2013            11A   
598737          8/17/2023            12D   
960725          12/1/2011           None   
366251           1/7/2022            10A   
24645            8/6/2007             8A   

                                    violation_description  \
692250  Manual facilities inadequate, technique incorr...   
598737  Toilet facilities inadequate, inconvenient, di...   
960725                                               None   
366251  Food (ice) contact surfaces are improperly des...   
24645   Food not protected during storage, preparation...   

            critical_violation  total_critical_violations  \
692250  Not Critical Violation                        0.0   
598737  Not Critical Violation                        0.0   
960725  Not Critical Violation                        0.0   
366251  Not Critical Violation                        0.0   
24645   Not Critical Violation                        0.0   

        total_crit_not_corrected  total_noncritical_violations  \
692250                       0.0                           4.0   
598737                       0.0                           3.0   
960725                       0.0                           0.0   
366251                       0.0                           2.0   
24645                        0.0                           5.0   

            local_health_department  ... permitted_dba  \
692250           Westchester County  ...          None   
598737            Livingston County  ...          None   
960725           Schenectady County  ...          None   
366251              Dutchess County  ...          None   
24645   Glens Falls District Office  ...          None   

               permitted_corp_name perm_operator_last_name  \
692250   LITTLE KABAB STATION INC.                   SARAN   
598737        Oasis Sweets and BBQ                    Howe   
960725     STADIUM GOLF CLUB, INC.                  HENNEL   
366251              18 NORTH GRILL                  INNESS   
24645   JBA RESTAURANT CORPORATION               ALEXANDER   

       perm_operator_first_name nys_health_operation_id  inspection_type  \
692250                   BONNIE                760285.0       Inspection   
598737                    Craig               1047213.0       Inspection   
960725                  GREGORY                305814.0       Inspection   
366251                    JAMEE               1058593.0       Inspection   
24645                    JASPER                337105.0       Inspection   

                                      inspection_comments fs_facility_state  \
692250                                               None                NY   
598737  All holding temps acceptable at time of inspec...                NY   
960725                                               None                NY   
366251  Adequate temperatures observed during inspecti...                NY   
24645                                                None                NY   

         latitude  longitude  
692250  41.208253 -73.727794  
598737  42.565351 -77.710524  
960725  42.780671 -73.914729  
366251        NaN        NaN  
24645   43.080425 -73.783635  

[5 rows x 31 columns]

In [15]:
insp.dtypes

facility                         object
address                          object
date_of_inspection               object
violation_item                   object
violation_description            object
critical_violation               object
total_critical_violations       float64
total_crit_not_corrected        float64
total_noncritical_violations    float64
local_health_department          object
county                           object
facility_code                    object
facility_address                 object
facility_city                    object
facility_postal_zipcode          object
nysdoh_gazetteer                float64
facility_municipality            object
operation_name                   object
permit_expiration_date           object
food_service_type                object
food_service_description         object
permitted_dba                    object
permitted_corp_name              object
perm_operator_last_name          object
perm_operator_first_name         object


- Everything looks good expect Columns`Date of inspection` , `Permit Expiration Date` , should be changed into datetime.

# 2.2 Duplicates

In [18]:
insp.duplicated().sum()

350

In [19]:
(
    insp
    .loc[insp.duplicated(keep=False)]
    .sort_values(['facility'])
)

facility                        address  \
1288367  2000 CHINESE RESTAURANT  22070 US ROUTE 11,  WATERTOWN   
484736   2000 CHINESE RESTAURANT  22070 US ROUTE 11,  WATERTOWN   
975267           4-H FAIRGROUNDS        12690 ROUTE 31,  ALBION   
1301839          4-H FAIRGROUNDS        12690 ROUTE 31,  ALBION   
129147           4-H FAIRGROUNDS        12690 ROUTE 31,  ALBION   
...                          ...                            ...   
1259115                     None                           None   
1260173                     None                           None   
1260857                     None                           None   
1262732                     None                           None   
1268373                     None                           None   

        date_of_inspection violation_item violation_description  \
1288367          6/10/2016           None                  None   
484736           6/10/2016           None                  None   
975267           7/23/2012           None                  None   
1301839          7/23/2012           None                  None   
129147           7/23/2012           None                  None   
...                    ...            ...                   ...   
1259115               None           None                  None   
1260173               None           None                  None   
1260857               None           None                  None   
1262732               None           None                  None   
1268373               None           None                  None   

             critical_violation  total_critical_violations  \
1288367  Not Critical Violation                        0.0   
484736   Not Critical Violation                        0.0   
975267   Not Critical Violation                        0.0   
1301839  Not Critical Violation                        0.0   
129147   Not Critical Violation                        0.0   
...                         ...                        ...   
1259115                    None                        NaN   
1260173                    None                        NaN   
1260857                    None                        NaN   
1262732                    None                        NaN   
1268373                    None                        NaN   

         total_crit_not_corrected  total_noncritical_violations  \
1288367                       0.0                           0.0   
484736                        0.0                           0.0   
975267                        0.0                           0.0   
1301839                       0.0                           0.0   
129147                        0.0                           0.0   
...                           ...                           ...   
1259115                       NaN                           NaN   
1260173                       NaN                           NaN   
1260857                       NaN                           NaN   
1262732                       NaN                           NaN   
1268373                       NaN                           NaN   

           local_health_department  ... permitted_dba  \
1288367  Watertown District Office  ...          None   
484736   Watertown District Office  ...          None   
975267              Orleans County  ...          None   
1301839             Orleans County  ...          None   
129147              Orleans County  ...          None   
...                            ...  ...           ...   
1259115                       None  ...          None   
1260173                       None  ...          None   
1260857                       None  ...          None   
1262732                       None  ...          None   
1268373                       None  ...          None   

                       permitted_corp_name perm_operator_last_name  \
1288367  2000 CHINESE RESTAURANT  NY, INC.                   Jiang   
484736   2000 CHINESE RESTAURANT  NY, I

- There are `350 Duplicates` in the dataset that should be removed

## 2.3 SHAPE OF THE DATASET

In [22]:
insp.shape

(1310873, 31)

- This is a huge dataset having `1.3 million` rows or entries and `31` columns.

## 3. DETAILED ANALYSIS

#### facility (Restaurant name)

In [26]:
insp.facility.unique()

array(['WENDYS OF EG', 'STEUBEN SR DINNER CLUB - BATH',
       'MOUNT KISCO COUNTRY CLUB - POOL SNAC', ..., 'COMMISSARY-LA SALVI',
       'SHALOM UMJCA', 'STARBUCKS (NYS THRUWAY)'], dtype=object)

In [27]:
insp.facility.nunique()

31971

In [28]:
insp.facility.value_counts()

facility
DUNKIN DONUTS                  5849
Dunkin Donuts                  2285
SUBWAY                         1993
CHINA GARDEN                   1252
DUNKIN' DONUTS                 1109
                               ... 
MANDY BEAN                        1
SCOOP & COFFEE                    1
AURORA BREWING CO                 1
WEST PATENT E.S. / WHITSONS       1
STARBUCKS (NYS THRUWAY)           1
Name: count, Length: 31971, dtype: int64

#### Address

In [30]:
insp.address.value_counts()

address
3349 MONROE AVENUE,  ROCHESTER              1244
1 CROSSGATES MALL ROAD,  ALBANY             1036
267 UNION AVENUE,  SARATOGA SPRINGS         1036
9993 Alleghany ROAD,  Darien Center          989
356 Villa Roma ROAD,  Callicoon              741
                                            ... 
159 DRYDEN ROAD,  ITHACA                       1
12 Reed STREET,  Coxsackie                     1
1026 WEST BEECH STREET,  LONG BEACH            1
195 WOODCLEFT AVENUE,  FREEPORT                1
NYS THRUWAY (M/M 153 EAST),  SCHENECTADY       1
Name: count, Length: 31989, dtype: int64

#### Violation food item

In [32]:
insp.violation_item.value_counts()

violation_item
15A    110164
8A      98253
10B     78069
11D     77756
15B     68602
        ...  
12          1
35          1
4           1
6           1
46          1
Name: count, Length: 98, dtype: int64

#### Inspection date

In [34]:
insp.date_of_inspection.value_counts()

date_of_inspection
11/15/2023    926
11/30/2023    846
1/31/2024     831
3/23/2023     831
2/1/2024      831
             ... 
10/28/2023      1
4/7/1997        1
9/20/2020       1
12/22/1995      1
7/6/1995        1
Name: count, Length: 9463, dtype: int64

In [35]:
pd.to_datetime(insp.date_of_inspection,dayfirst = True,format='mixed')

OutOfBoundsDatetime: Out of bounds nanosecond timestamp: 06/20/0200, at position 2565

In [49]:
(
    insp.
    date_of_inspection.
    loc[2565]
)

'06/20/0200'

In [51]:
# Fixing in correct data(year) at position 2565
insp.loc[insp['date_of_inspection'] == '06/20/0200', 'date_of_inspection'] = '06/20/2000'

In [53]:
invalid_dates = insp[pd.to_datetime(insp['date_of_inspection'], errors='coerce').isnull()]['date_of_inspection']

print(invalid_dates)

30               None
2896       04/07/0005
5640       07/14/0004
46138            None
119759           None
              ...    
1287453    12/08/0002
1291687    02/20/0300
1298900    12/02/0003
1307693    09/08/0300
1310690    04/21/0300
Name: date_of_inspection, Length: 87, dtype: object


In [55]:
invalid_dates.value_counts()

date_of_inspection
04/30/0200    3
01/20/0300    2
11/03/0005    2
04/07/0005    1
07/30/0005    1
01/23/0002    1
04/17/0003    1
04/25/0600    1
01/05/1200    1
08/04/0022    1
07/04/0022    1
10/26/1001    1
02/24/0060    1
03/14/0200    1
02/04/0020    1
05/25/0200    1
01/14/0031    1
05/16/1200    1
12/08/0002    1
02/20/0300    1
12/02/0003    1
09/08/0300    1
01/14/0040    1
01/10/0184    1
11/18/0300    1
07/14/0004    1
09/19/0003    1
11/18/0505    1
03/24/0300    1
11/13/0500    1
03/23/0006    1
04/04/0022    1
02/21/0200    1
01/17/0605    1
11/17/0300    1
02/04/1003    1
03/19/0403    1
12/04/1002    1
10/14/0021    1
06/24/0005    1
11/23/1005    1
01/05/1002    1
01/29/0300    1
04/21/0300    1
Name: count, dtype: int64

#### Critical Violation 

In [57]:
insp.critical_violation.value_counts()

critical_violation
Not Critical Violation    1158252
Critical Violation         152582
Name: count, dtype: int64

#### total_critical_violations, total_crit_not_corrected, total_noncritical_violations

In [59]:
insp[['total_critical_violations', 'total_crit_not_corrected', 'total_noncritical_violations']].describe()

total_critical_violations  total_crit_not_corrected  \
count               1.230072e+06              1.229454e+06   
mean                8.127939e-01              2.876073e-02   
std                 1.436214e+00              2.947295e-01   
min                 0.000000e+00              0.000000e+00   
25%                 0.000000e+00              0.000000e+00   
50%                 0.000000e+00              0.000000e+00   
75%                 1.000000e+00              0.000000e+00   
max                 8.000000e+01              8.800000e+01   

       total_noncritical_violations  
count                  1.230201e+06  
mean                   5.250862e+00  
std                    5.160587e+00  
min                    0.000000e+00  
25%                    2.000000e+00  
50%                    4.000000e+00  
75%                    7.000000e+00  
max                    9.800000e+01

- Avearge  `violations` for restaurants is around `6.06`
- In those ,`Critical - 0.81`  , `Non-Critical-5.25`
- Average `critical cases` that are not corrected - `00.287` which is pretty low

#### local_health_department 

In [61]:
insp.local_health_department.value_counts()

local_health_department
Westchester County                                              175883
Nassau  County                                                  160850
Monroe County                                                   117206
Onondaga County                                                  79448
Glens Falls District Office                                      53669
Dutchess County                                                  52625
Rockland County                                                  47524
Orange County                                                    46475
Broome County                                                    44638
Niagara County                                                   38557
Albany County                                                    38247
Saranac Lake District Office                                     35425
Oneonta District Office                                          32182
Ulster County                                        

#### Facility muncipality

In [63]:
insp.facility_municipality.value_counts()


facility_municipality
HEMPSTEAD          76854
ROCHESTER          42360
N. HEMPSTEAD       35105
OYSTER BAY         33497
SYRACUSE           28710
                   ...  
CATON                  1
AMBOY                  1
WETHERSFIELD           1
NASSAU COUNTY          1
SCHUYLER COUNTY        1
Name: count, Length: 914, dtype: int64

#### Permit Expiry date

In [65]:
insp.permit_expiration_date.value_counts()

permit_expiration_date
12/31/2024    131697
3/31/2025     127146
1/31/2025     126668
9/30/2024     110679
12/31/2019     98571
               ...  
4/30/2020          1
10/10/2024         1
8/13/2021          1
10/30/2022         1
12/27/2024         1
Name: count, Length: 404, dtype: int64

In [67]:
pd.to_datetime(insp.permit_expiration_date,dayfirst=True,format='mixed')

OutOfBoundsDatetime: Out of bounds nanosecond timestamp: 11/30/3018, at position 271

In [69]:
invalid_permit_dates = insp[pd.to_datetime(insp['permit_expiration_date'], errors='coerce').isnull()]['permit_expiration_date']


In [71]:
invalid_permit_dates.value_counts()

permit_expiration_date
3/31/3201     30
11/30/3018    17
9/30/3024      3
12/30/2400     2
Name: count, dtype: int64

#### food service type 

In [73]:
insp.food_service_type.value_counts()

food_service_type
Food Service Establishment    1131871
Institutional Food Service     154703
SED Summer Feeding Prog.        14210
SOFA Food Service               10050
Name: count, dtype: int64

#### Inspection Type

In [75]:
insp.inspection_type.value_counts()

inspection_type
Inspection       1188410
Re-Inspection     122424
Name: count, dtype: int64

## 3.1 NULL VALUES

In [77]:
insp.isna().sum()

facility                             39
address                              41
date_of_inspection                   39
violation_item                   241221
violation_description            241221
critical_violation                   39
total_critical_violations         80801
total_crit_not_corrected          81419
total_noncritical_violations      80672
local_health_department              39
county                               39
facility_code                        39
facility_address                     45
facility_city                       172
facility_postal_zipcode             172
nysdoh_gazetteer                     39
facility_municipality                39
operation_name                       39
permit_expiration_date            15672
food_service_type                    39
food_service_description             39
permitted_dba                   1177058
permitted_corp_name              305477
perm_operator_last_name          334093
perm_operator_first_name         335787


## 4. CLEANING DATA

- Imputing null values for columns `'total_critical_violations', 'total_crit_not_corrected',
       'total_noncritical_violations'` using `MICE` Algorithm

In [79]:
from sklearn.experimental import enable_iterative_imputer
from sklearn.impute import IterativeImputer

In [81]:
cols =['total_critical_violations', 'total_crit_not_corrected',
       'total_noncritical_violations']

In [83]:

invalid_dates = [x for x in invalid_dates if x is not None]

invalid_permit_dates = [x for x in invalid_permit_dates if x is not None]
def clean_data(df,cols):
    imputer = IterativeImputer(random_state=42)
    df[cols] = imputer.fit_transform(df[cols])
    df = df[~df['permit_expiration_date'].isin(invalid_permit_dates)]
    df = df[~df['date_of_inspection'].isin(invalid_dates)]
    return(
        df
    )

In [85]:
invalid_dates

['04/07/0005',
 '07/14/0004',
 '09/19/0003',
 '04/30/0200',
 '11/18/0505',
 '03/24/0300',
 '11/13/0500',
 '03/23/0006',
 '04/04/0022',
 '02/21/0200',
 '01/17/0605',
 '11/17/0300',
 '02/04/1003',
 '03/19/0403',
 '12/04/1002',
 '11/03/0005',
 '10/14/0021',
 '06/24/0005',
 '11/23/1005',
 '01/05/1002',
 '01/20/0300',
 '04/30/0200',
 '01/29/0300',
 '01/20/0300',
 '01/10/0184',
 '11/18/0300',
 '03/14/0200',
 '01/14/0040',
 '01/23/0002',
 '04/17/0003',
 '04/30/0200',
 '04/25/0600',
 '01/05/1200',
 '08/04/0022',
 '07/04/0022',
 '10/26/1001',
 '11/03/0005',
 '07/30/0005',
 '02/24/0060',
 '02/04/0020',
 '05/25/0200',
 '01/14/0031',
 '05/16/1200',
 '12/08/0002',
 '02/20/0300',
 '12/02/0003',
 '09/08/0300',
 '04/21/0300']

In [87]:
invalid_permit_dates

['11/30/3018',
 '3/31/3201',
 '3/31/3201',
 '3/31/3201',
 '3/31/3201',
 '11/30/3018',
 '3/31/3201',
 '11/30/3018',
 '11/30/3018',
 '11/30/3018',
 '3/31/3201',
 '11/30/3018',
 '11/30/3018',
 '3/31/3201',
 '3/31/3201',
 '3/31/3201',
 '3/31/3201',
 '11/30/3018',
 '3/31/3201',
 '3/31/3201',
 '3/31/3201',
 '11/30/3018',
 '3/31/3201',
 '3/31/3201',
 '11/30/3018',
 '3/31/3201',
 '3/31/3201',
 '3/31/3201',
 '11/30/3018',
 '3/31/3201',
 '11/30/3018',
 '3/31/3201',
 '11/30/3018',
 '3/31/3201',
 '9/30/3024',
 '12/30/2400',
 '3/31/3201',
 '3/31/3201',
 '11/30/3018',
 '11/30/3018',
 '3/31/3201',
 '3/31/3201',
 '11/30/3018',
 '3/31/3201',
 '3/31/3201',
 '9/30/3024',
 '3/31/3201',
 '3/31/3201',
 '3/31/3201',
 '9/30/3024',
 '11/30/3018',
 '12/30/2400']

In [89]:
print(len(invalid_dates),len(invalid_permit_dates))

48 52


In [91]:
clean_insp = clean_data(insp,cols)

In [93]:
clean_insp['total_noncritical_violations'].isna().sum()

0

In [95]:
clean_insp.shape

(1310773, 31)

- Null values in `Numerical` Columns `['total_critical_violations', 'total_crit_not_corrected',
       'total_noncritical_violations']` are imputed using `mice` Algorithm.
- `Incorrect Date Entries` columns are removed.

- Null values in `Categorical Columns` are imputed after checking their distribution and impute method is selcted based on their distribution or skewness.

## END OF INTIAL INSPECTION AND DATACLEANING